client/pluginmanager/drivermanager/manager.go

```diff
@@ -0,0 +1,286 @@
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// loader is the plugin loader
+	loader loader.PluginCatalog
+
+	// pluginConfig is the config passed to the launched plugins
+	pluginConfig *base.ClientAgentConfig
+
+	// updater is used to update the node when device information changes
+	updater UpdateNodeDriverInfoFn
+
+	// instances is the list of managed devices
+	instances map[string]*instanceManager
+
+	// reattachConfigs stores the plugin reattach configs
+	reattachConfigs    map[loader.PluginID]*shared.ReattachConfig
+	reattachConfigLock sync.Mutex
+
+	// allows/block lists
+	allowedDrivers map[string]struct{}
+	blockedDrivers map[string]struct{}
+}
+
+// New returns a new driver manager
+func New(c *Config) *manager {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &manager{
+		logger:          c.Logger.Named("driver_mgr"),
+		state:           c.State,
+		ctx:             ctx,
+		cancel:          cancel,
+		loader:          c.Loader,
+		pluginConfig:    c.PluginConfig,
+		updater:         c.Updater,
+		instances:       make(map[string]*instanceManager),
+		reattachConfigs: make(map[loader.PluginID]*shared.ReattachConfig),
+		allowedDrivers:  c.AllowedDrivers,
+		blockedDrivers:  c.BlockedDrivers,
+	}
+}
+
+// PluginType returns the type of plugin this mananger mananges
+func (*manager) PluginType() string { return base.PluginTypeDriver }
+
+// Run starts the mananger, initializes driver plugins and blocks until Shutdown
+// is called.
+func (m *manager) Run() {
+	// Load any previous plugin reattach configuration
+	m.loadReattachConfigs()
+
+	// Get driver plugins
+	driversPlugins := m.loader.Catalog()[base.PluginTypeDriver]
+	if len(driversPlugins) == 0 {
+		m.logger.Debug("exiting since there are no driver plugins")
+		m.cancel()
+		return
+	}
+
+	var availDrivers []string
+	var skippedDrivers []string
+	for _, d := range driversPlugins {
+		id := loader.PluginInfoID(d)
+		// Skip drivers that are not in the allowed list if it is set.
+		if _, ok := m.allowedDrivers[id.Name]; len(m.allowedDrivers) > 0 && !ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+		// Skip fingerprinting drivers that are in the blocked list
+		if _, ok := m.blockedDrivers[id.Name]; ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+
+		storeFn := func(c *plugin.ReattachConfig) error {
+			id := id
+			return m.storePluginReattachConfig(id, c)
+		}
+		fetchFn := func() (*plugin.ReattachConfig, bool) {
+			id := id
+			return m.fetchPluginReattachConfig(id)
+		}
+
+		instance := newInstanceManager(&instanceManagerConfig{
+			Logger:               m.logger,
+			Ctx:                  m.ctx,
+			Loader:               m.loader,
+			StoreReattach:        storeFn,
+			FetchReattach:        fetchFn,
+			PluginConfig:         m.pluginConfig,
+			ID:                   &id,
+			UpdateNodeFromDriver: m.updater,
+		})
+
+		ctx, cancel := context.WithTimeout(m.ctx, 1*time.Second)
```

Comment:
Is 1 second sufficient? Let's comment explaining the choice.

Reply:
So the only purpose of this is to gather a list of "detected" drivers and log them. The timeout doesn't affect fingerprinting. 

It was 5 seconds and ideally should be pretty instant since drivers immediately send a fingerprint when the RPC is started. However on travis, with CPU contention and the high number of goroutines running it can take longer and since launching the driver manager is async, it was causing some really wierd test conditions because the drivers weren't loaded before the tr dispensed the driver.

I'm tempted to remove this entirely, however it is a really helpful log message when running nomad, especially in dev mode. Thoughts?

Reply:
Ah, if it's just for logging why not bump the timeout up higher just to be safe? My virtualbox vm isn't exactly speedy. :)

Also I think it's better to be late logging than to timeout on a slow-but-healthy driver, have it miss the log line, and make the user believe something is wrong.

nbd either way though since this is just for logging. Not a blocker.